docs/rfcs/2025-07-07-node-deletion-api-improvement.md

```diff
@@ -0,0 +1,149 @@
+deployment operations.
+
+### Reliability, failure modes and corner cases
+
+In case of a storage controller failure and following restart, the system behavior depends on the
+`NodeLifecycle` state:
+
+- If `NodeLifecycle` is `Active`: No action is taken for this node.
+- If `NodeLifecycle` is `Deleted`: The node will not be re-added.
+- If `NodeLifecycle` is `ScheduledForDeletion`: A deletion background task will be launched for
+this node. Since only one ongoing deletion task per storage controller is allowed, if there are
+multiple nodes with `ScheduledForDeletion` state, additional deletion tasks will be queued.
+
+In case of a pageserver node failure during deletion, the behavior depends on the `force` flag:
+- If `force` is set: The node deletion will proceed regardless of the node's availability.
+- If `force` is not set: The deletion will be cancelled after some number of retry attempts.
+
+### Operations concurrency
+
+The following sections describe the behavior when different types of requests arrive at the storage
+controller and how they interact with ongoing operations.
+
+#### Delete request
+
+1. If node lifecycle is `NodeLifecycle::ScheduledForDeletion`:
+    - Return `409 Conflict`: there is already an ongoing deletion request for this node
+2. Update & persist lifecycle to `NodeLifecycle::ScheduledForDeletion`
+3. If there is no active operation (drain/fill/delete):
+    - Run deletion process for this node
+
+#### Cancel delete request
+
+1. If node lifecycle is not `NodeLifecycle::ScheduledForDeletion`:
+    - Return `404 Not Found`: there is no current deletion request for this node
+2. If the active operation is deleting this node, cancel it
+3. Update & persist lifecycle to `NodeLifecycle::Active`
+4. If there is no active operation (drain/fill/delete):
+    - Try to find another candidate to delete and run the deletion process for that node
+
+#### Drain/fill request
+
+1. If there are already ongoing drain/fill processes:
+    - Return `409 Conflict`: queueing of drain/fill processes is not supported
+2. If there is an ongoing delete process:
+    - Cancel it and wait until it is cancelled
+3. Run the drain/fill process
+4. After the drain/fill process is cancelled or finished:
+    - Try to find another candidate to delete and run the deletion process for that node
+
+#### Drain/fill cancel request
+
+1. If the active operation is not the related process:
+    - Return `400 Bad Request`: cancellation request is incorrect, operations are not the same
+2. Cancel the active operation
+3. Try to find another candidate to delete and run the deletion process for that node
+
+### Unresolved questions
+
+- Should fill operations be treated differently than drain operations when processing deletion? For
+example, should a fill operation be cancelled immediately when a deletion is requested, while drain
+operations might be allowed to complete first?
```

Comment:
We should allow the fill to complete. Filling moves back shards into the right AZ, so it matters for performancce. The goal of the graceful deletion is to preseve that.